app/controllers/api/v2/hosts_bulk_actions_controller.rb

```diff
@@ -78,6 +78,38 @@ def reassign_hostgroup
         end
       end
 
+      api :PUT, "/hosts/bulk/assign_taxonomy", N_("Assign organization/location")
+      param_group :bulk_host_ids
+      param :target_organization_id, :number, :desc => N_("ID of the organization to assign the hosts to")
+      param :target_location_id, :number, :desc => N_("ID of the location to assign the hosts to")
+      param :optimistic_import_organization, :bool, N_("Fix organization on mismatch")
+      param :optimistic_import_location, :bool, N_("Fix location on mismatch")
+      def assign_taxonomy
+        if [params[:optimistic_import_organization], params[:optimistic_import_location]].any? { |val| !val }
+          render_error(:custom_error, :status => :unprocessable_entity, :locals => { :message => _("Cannot update because of mismatch in settings.")})
+          return
+        end
+
+        begin
+          clear_current_taxonomy
+          find_editable_hosts
+
+          messages = []
+          targets = Taxonomy.where(id: [params[:target_organization_id], params[:target_location_id]])
+          targets.each do |tax|
+            tax_type = tax.type.downcase
+            BulkHostsManager.new(hosts: @hosts).assign_taxonomy(tax, params["optimistic_import_#{tax_type}"])
+            msg = tax_type == 'organization' ? _("Organization is set to %s.") % tax.name : _("Location is set to %s.") % tax.name
+            messages << msg
+          end
```

Comment:
Nit: you can shorten this
```suggestion
          targets = Taxonomy.where(id: [params[:target_organization_id], params[:target_location_id]])
          messages = targets.map do |tax|
            tax_type = tax.type.downcase
            BulkHostsManager.new(hosts: @hosts).assign_taxonomy(tax, params["optimistic_import_#{tax_type}"])
            tax_type == 'organization' ? _("Organization is set to %s.") % tax.name : _("Location is set to %s.") % tax.name
          end
```